File: src/main/java/szewek/mctool/app/task/TaskManager.java
package szewek.mctool.app.task;

import javafx.application.Platform;
import javafx.beans.binding.Bindings;
import javafx.beans.binding.ObjectBinding;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.concurrent.Worker;

public class TaskManager {
    private TaskManager() {}

    public static final BooleanProperty taskListOpen = new SimpleBooleanProperty(false);
    public static final ObservableList<Task<?>> tasks = FXCollections.observableArrayList();
    public static final ObjectBinding<Task<?>> lastTask = Bindings.createObjectBinding(TaskManager::findTask, tasks);

    public static void addTask(final Task<?> t) {
        if (!Platform.isFxApplicationThread()) {
            Platform.runLater(() -> addTask(t));
        }
        System.out.println("ADDING TASK " + t);
        tasks.add(t);
        removeOnFinish(t);
    }

    private static void removeOnFinish(final Task<?> t) {
        var csl = new ChangeListener<Worker.State>() {
            @Override
            public void changed(ObservableValue<? extends Worker.State> ov, Worker.State os, Worker.State ns) {
                if (ns == Worker.State.SUCCEEDED || ns == Worker.State.CANCELLED || ns == Worker.State.FAILED) {
                    tasks.remove(t);
                    ov.removeListener(this);
                }
            }
        };
        t.stateProperty().addListener(csl);
    }

    private static Task<?> findTask() {
        var iter = tasks.listIterator(tasks.size());
        while (iter.hasPrevious()) {
            var t = iter.previous();
            if (!t.isDone()) {
                return t;
            }
        }
        return null;
    }
}

File: src/main/kotlin/szewek/mctool/app/task/TaskManager.kt
package szewek.mctool.app.task

import javafx.application.Platform
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.concurrent.Task
import javafx.concurrent.Worker
import szewek.mctool.app.objectBinding
import tornadofx.FXTask
import tornadofx.finally
import tornadofx.observableListOf
import tornadofx.task
import java.util.*
import java.util.concurrent.ArrayBlockingQueue

object TaskManager {
    val taskListOpen = SimpleBooleanProperty(false)
    val tasks = observableListOf<Task<*>>()
    val lastTask = tasks.objectBinding { findLast { !it.isDone } }

    fun addTask(t: Task<*>) {
        if (Platform.isFxApplicationThread()) {
            tasks += t
            removeOnFinish(t)
        } else {
            Platform.runLater {
                tasks += t
                removeOnFinish(t)
            }
        }
    }

    private fun removeOnFinish(t: Task<*>) {
        val csl = object : ChangeListener<Worker.State> {
            override fun changed(ov: ObservableValue<out Worker.State>, os: Worker.State?, ns: Worker.State?) {
                if (ns == Worker.State.SUCCEEDED || ns == Worker.State.CANCELLED || ns == Worker.State.FAILED) {
                    tasks.remove(t)
                    ov.removeListener(this)
                }
            }
        }
        t.stateProperty().addListener(csl)
    }

    fun createTaskQueue(vararg tt: FXTask<*>.() -> Unit) {
        val q = ArrayBlockingQueue<FXTask<*>.() -> Unit>(tt.size)
        q.addAll(tt)
        genTask(q)
    }

    private fun genTask(tt: TaskQueue) {
        val fn = tt.poll()
        if (fn != null) {
            addTask(task(func = fn).apply { finally { genTask(tt) } })
        }
    }
}

typealias TaskFunc = FXTask<*>.() -> Unit
typealias TaskQueue = Queue<TaskFunc>